files/json_output_formatter.rb

```diff
@@ -15,19 +15,24 @@ def dump_summary(summary)
       map { |example| example.metadata[:points].to_i }.
       sum
 
+    score = (earned_points.to_f / total_points).round(4)
+    if score.nan?
+      score = 0
+    end
+    
     @output_hash[:summary] = {
       duration: summary.duration,
       example_count: summary.example_count,
       failure_count: summary.failure_count,
       pending_count: summary.pending_count,
       total_points: total_points,
       earned_points: earned_points,
-      score: (earned_points.to_f / total_points).round(4)
+      score: score
     }
     score = (@output_hash[:summary][:score] * 100).round(2)
     
-    if score.nan?
-      score = "This project is not graded."
+    if summary.example_count.zero?
```

Comment:
Is it necessarily true that an error occurred if there were `0` tests? Could it be that we just didn't include any tests, in which case the message should be "There were no tests in this project."?

Reply:
With how all projects and grades is set up, when a assignment is not worth points:
1. has a `dummy_spec` file that has the message, `"This project is not graded"`
2. the grades page hides the token so they _can't_ even run `rails grade`

I believe the only other time they could get this error is when there is an error in the code that prevents `rspec` from completing. This also allows for tests that are worth `0` points (like what we tried to do with `a-quick-puzzle`) to not break the formatter.

Reply:
I hear you but I wonder if it would be more robust not to depend upon projects being correctly set up. It also seems like this sanitization is happening too late, _after_ score is being used in math and placed in the output hash. But it's been a long time since I've looked at the code so I could be misunderstanding the order of operations.

Instead, it seems to me like the logic should be something like,

 - Are there 0 examples? Then set `score` to `100`,  `earned_points` and `total_points` to `0`, `result` to `:not_graded`, and `:message` to `"This project is not graded."`.
 - Else, calculate the score as before. Set `result` to `:graded`, `:message` to "#{score}%".
 - `if score.nan?` then update `score` to `0`,  `earned_points` and `total_points` to `0`, `result` to `:error`, and `message` to `"Your code caused an error while the tests were running, which prevented the test suite from completing."`.

Then we can finally use a `case` statement on `result` to craft the `@output_hash`. What do you think?

Reply:
I don't understand what `result` is in your example. Would modifying the structure of this format require a change in grades?

> Are there 0 examples? Then set score to 1.0, earned_points and total_points to 0, result to :not_graded, and :message to "This project is not graded.".

If we're operating under the assumption that some projects will not be set up correctly, counting the number of examples is not enough information to determine the score or points received. Errors that interrupt `rspec` will have an example count of `0`. See https://github.com/firstdraft/appdev_template/pull/156#discussion_r457451042

`score.nan?` doesn't necessarily solve the infinity problem by itself
```ruby
Float::INFINITY.nan? # => false
```

Reply:
`result` is just a random variable I made up to capture the status of the build. Maybe `status` would have been a better name.

The point is, now that we have more cases, I think we should capture what's happening and then probably have a `case` statement at the end to put together the final output, rather than massaging the same variable along the way.

Reply:
> Would modifying the structure of this format require a change in grades?

I don't mean to modify the structure of the final format at all; just how we arrive at it. But no big deal, your original proposed solution gets the job done.